lib/railway_ipc/core/payload.ex

```diff
@@ -1,53 +1,37 @@
 defmodule RailwayIpc.Core.Payload do
-  @moduledoc false
-  @behaviour RailwayIpc.PayloadBehaviour
-  import RailwayIpc.Utils, only: [module_defined?: 1]
+  @moduledoc """
+  _This is an internal module, not part of the public API._
 
-  alias RailwayIpc.DefaultMessage
+  Handles encoding/decoding of messages by delegating to specific message
+  format helper modules.
 
-  def decode(payload) when not is_binary(payload) do
-    {:error, "Malformed JSON given: #{payload}. Must be a string"}
-  end
-
-  def decode(payload) do
-    with {:decode_json, {:ok, %{"type" => type, "encoded_message" => encoded_message}}} <- {
-           :decode_json,
-           Jason.decode(payload)
-         },
-         {:convert_module, module} <- {:convert_module, module_from_type(type)},
-         {:check_module_exists, true} <- {:check_module_exists, module_defined?(module)},
-         {:decode_message, message} <- {:decode_message, decode_message(module, encoded_message)} do
-      {:ok, message}
-    else
-      {:decode_json, {:ok, _}} ->
-        {:error, "Missing keys: #{payload}. Expecting type and encoded_message keys"}
+  """
 
-      {:decode_json, {:error, _}} ->
-        {:error, "Malformed JSON given: #{payload}"}
+  alias RailwayIpc.Core.MessageFormat.BinaryProtobuf
 
-      {:check_module_exists, false} ->
-        %{"encoded_message" => encoded_message, "type" => type} = Jason.decode!(payload)
-        {:unknown_message_type, decode_message(DefaultMessage, encoded_message), type}
+  def decode(payload) do
+    case BinaryProtobuf.decode(payload) do
+      # FIXME: We should be consistent and return type for the :ok case
```

Comment:
Future work planned?

Reply:
yeah, future work; too much to refactor so I want to do it in smaller pieces